File: week-2/number-of-atoms/number-of-atoms.js
/**
 * @param {string} formula
 * @return {string}
 */
var countOfAtoms = function(formula) {
    const hashtable = {};
    let power = 0, count = 0, coefficient = 1;
    let elem = '';
    const stack = [];
    
    for (let i of formula.split('').reverse()) {
        if (i >= '0' && i<='9') {
            count += parseInt(i) * (10**power);
            ++power;
        } else if (i === '(') {
            coefficient /= stack.pop();
            count = power = 0;
        } else if (i === ')') {
            stack.push(count);
            coefficient *= count;
            count = power = 0;
        } else if (i === i.toLowerCase()) {
            elem += i;
        } else if (i === i.toUpperCase()) {
            elem += i;
            rev = elem.split('').reverse().join('');
            if (!(rev in hashtable)) hashtable[rev] = 0;
            hashtable[rev] += (count || 1) * coefficient;
            count = power = 0;
            elem = '';
        }
    }
        
    return Object.entries(hashtable).sort().map(item => item[0] + (item[1] === 1 ? '' : item[1])).join('');
    
};

File: week-2/number-of-atoms/number-of-atoms.py
class Solution:
    def countOfAtoms(self, formula: str) -> str:
        hashtable = collections.defaultdict(int)
        stack = []
        elem = ''
        count = 0
        power = 0
        coefficient = 1
        for i in formula[::-1]:
            if i.islower():
                elem += i
            
            elif i.isupper():
                elem += i
                hashtable[elem[::-1]] += (count or 1) * coefficient
                count = power = 0
                elem = ''
            
            elif i.isdigit():
                count += int(i) * (10**power)
                power += 1
            
            elif i == ')':
                stack.append(count)
                coefficient *= count
                count = power = 0
                
            elif i == '(':
                coefficient //= stack.pop()
                count = power = 0
                
        x = list(hashtable.items())
        x.sort(key=lambda a: a[0])
        return ''.join(list(map(lambda a: a[0] + str(a[1]) if a[1] != 1 else a[0], x)))
